### ISSUE TITLE
RuntimeError: expected HostSystem for 'host', got array

### ISSUE THREAD

**Issue by g0atcheez** [CenturyLinkCloud/chef-provisioning-vsphere#4]
Hello,

I'm running into a roadblock whenever trying to test out provisioning using vCenter in my home lab. It's probably something I'm doing wrong, but I thought I'd see if someone can help.

ChefDK: 0.6.0
vCenter 6.0

There are no clusters or resource pools setup, thus I'm trying to utilize the :host parameter. Could be because I'm running vCenter 6.0, but wanted to confirm before doing any downgrading or rebuilding. Thanks in advance.

Error message:

> [2015-06-10T21:18:36-05:00] FATAL: RuntimeError: machine [my_machine_name] (vcenter_provisioning_test::provision line 42) had an error: RuntimeError: expected HostSystem for 'host', got array

::provision recipe
```ruby
chef_gem 'chef-provisioning-vsphere' do
  action :install
  compile_time true
end

require 'chef/provisioning/vsphere_driver'

with_vsphere_driver host: '192.168.1.112',
  insecure: true,
  user:     '****',
  password: '****'

with_machine_options :bootstrap_options => {
  use_linked_clone: true,
  num_cpus: 1,
  memory_mb: 4096,
  network_name: ['VM Network'],
  datacenter: 'Home',
  host: '192.168.1.100',
  datastore: 'esxlab_storage',
  template_name: 'WIN2008R2PS4',
  customization_spec: {
    ipsettings: {
      dnsServerList: ['192.168.1.110','192.168.1.1']
    },
    domain: 'casa.local',
    domainAdmin: 'administrator@casa.local',
    domainAdminPassword: '****',
    org_name: 'Casa',
    win_time_zone: 20
  },
  ssh: {
    user: 'Administrator',
    password: '****',
    paranoid: false,
  }
},
:convergence_options => {
  :ssl_verify_mode => :verify_none
}

machine "my_machine_name" do
  run_list ['vcenter_provisioning_test::default']
end
```

Stack trace
> Generated at 2015-06-10 21:18:36 -0500
RuntimeError: machine[my_machine_name] (vcenter_provisioning_test::provision line 42) had an error: RuntimeError: expected HostSystem for 'host', got array
C:/Users/Matthew/.chefdk/gem/ruby/2.1.0/gems/rbvmomi-1.8.2/lib/rbvmomi/connection.rb:129:in `obj2xml'
C:/Users/Matthew/.chefdk/gem/ruby/2.1.0/gems/rbvmomi-1.8.2/lib/rbvmomi/connection.rb:146:in `block (2 levels) in obj2xml'
C:/Users/Matthew/.chefdk/gem/ruby/2.1.0/gems/rbvmomi-1.8.2/lib/rbvmomi/connection.rb:142:in `each'
C:/Users/Matthew/.chefdk/gem/ruby/2.1.0/gems/rbvmomi-1.8.2/lib/rbvmomi/connection.rb:142:in `block in obj2xml'
C:/opscode/chefdk/embedded/lib/ruby/gems/2.1.0/gems/builder-3.2.2/lib/builder/xmlbase.rb:175:in `call'
C:/opscode/chefdk/embedded/lib/ruby/gems/2.1.0/gems/builder-3.2.2/lib/builder/xmlbase.rb:175:in `_nested_structures'
C:/opscode/chefdk/embedded/lib/ruby/gems/2.1.0/gems/builder-3.2.2/lib/builder/xmlbase.rb:68:in `tag!'
C:/Users/Matthew/.chefdk/gem/ruby/2.1.0/gems/rbvmomi-1.8.2/lib/rbvmomi/connection.rb:141:in `obj2xml'
C:/Users/Matthew/.chefdk/gem/ruby/2.1.0/gems/rbvmomi-1.8.2/lib/rbvmomi/connection.rb:146:in `block (2 levels) in obj2xml'
C:/Users/Matthew/.chefdk/gem/ruby/2.1.0/gems/rbvmomi-1.8.2/lib/rbvmomi/connection.rb:142:in `each'
C:/Users/Matthew/.chefdk/gem/ruby/2.1.0/gems/rbvmomi-1.8.2/lib/rbvmomi/connection.rb:142:in `block in obj2xml'
C:/opscode/chefdk/embedded/lib/ruby/gems/2.1.0/gems/builder-3.2.2/lib/builder/xmlbase.rb:175:in `call'
C:/opscode/chefdk/embedded/lib/ruby/gems/2.1.0/gems/builder-3.2.2/lib/builder/xmlbase.rb:175:in `_nested_structures'
C:/opscode/chefdk/embedded/lib/ruby/gems/2.1.0/gems/builder-3.2.2/lib/builder/xmlbase.rb:68:in `tag!'
C:/Users/Matthew/.chefdk/gem/ruby/2.1.0/gems/rbvmomi-1.8.2/lib/rbvmomi/connection.rb:141:in `obj2xml'
C:/Users/Matthew/.chefdk/gem/ruby/2.1.0/gems/rbvmomi-1.8.2/lib/rbvmomi/connection.rb:47:in `block (2 levels) in emit_request'
C:/Users/Matthew/.chefdk/gem/ruby/2.1.0/gems/rbvmomi-1.8.2/lib/rbvmomi/connection.rb:42:in `each'
C:/Users/Matthew/.chefdk/gem/ruby/2.1.0/gems/rbvmomi-1.8.2/lib/rbvmomi/connection.rb:42:in `block in emit_request'
C:/opscode/chefdk/embedded/lib/ruby/gems/2.1.0/gems/builder-3.2.2/lib/builder/xmlbase.rb:175:in `call'
C:/opscode/chefdk/embedded/lib/ruby/gems/2.1.0/gems/builder-3.2.2/lib/builder/xmlbase.rb:175:in `_nested_structures'
C:/opscode/chefdk/embedded/lib/ruby/gems/2.1.0/gems/builder-3.2.2/lib/builder/xmlbase.rb:68:in `tag!'
C:/Users/Matthew/.chefdk/gem/ruby/2.1.0/gems/rbvmomi-1.8.2/lib/rbvmomi/connection.rb:40:in `emit_request'
C:/Users/Matthew/.chefdk/gem/ruby/2.1.0/gems/rbvmomi-1.8.2/lib/rbvmomi/connection.rb:83:in `block in call'
C:/Users/Matthew/.chefdk/gem/ruby/2.1.0/gems/rbvmomi-1.8.2/lib/rbvmomi/trivial_soap.rb:71:in `block (2 levels) in soap_envelope'
C:/opscode/chefdk/embedded/lib/ruby/gems/2.1.0/gems/builder-3.2.2/lib/builder/xmlbase.rb:175:in `call'
C:/opscode/chefdk/embedded/lib/ruby/gems/2.1.0/gems/builder-3.2.2/lib/builder/xmlbase.rb:175:in `_nested_structures'
C:/opscode/chefdk/embedded/lib/ruby/gems/2.1.0/gems/builder-3.2.2/lib/builder/xmlbase.rb:68:in `tag!'
C:/Users/Matthew/.chefdk/gem/ruby/2.1.0/gems/rbvmomi-1.8.2/lib/rbvmomi/trivial_soap.rb:70:in `block in soap_envelope'
C:/opscode/chefdk/embedded/lib/ruby/gems/2.1.0/gems/builder-3.2.2/lib/builder/xmlbase.rb:175:in `call'
C:/opscode/chefdk/embedded/lib/ruby/gems/2.1.0/gems/builder-3.2.2/lib/builder/xmlbase.rb:175:in `_nested_structures'
C:/opscode/chefdk/embedded/lib/ruby/gems/2.1.0/gems/builder-3.2.2/lib/builder/xmlbase.rb:68:in `tag!'
C:/Users/Matthew/.chefdk/gem/ruby/2.1.0/gems/rbvmomi-1.8.2/lib/rbvmomi/trivial_soap.rb:64:in `soap_envelope'
C:/Users/Matthew/.chefdk/gem/ruby/2.1.0/gems/rbvmomi-1.8.2/lib/rbvmomi/connection.rb:82:in `call'
C:/Users/Matthew/.chefdk/gem/ruby/2.1.0/gems/rbvmomi-1.8.2/lib/rbvmomi/basic_types.rb:205:in `_call'
C:/Users/Matthew/.chefdk/gem/ruby/2.1.0/gems/rbvmomi-1.8.2/lib/rbvmomi/basic_types.rb:74:in `block (2 levels) in init'
C:/Users/Matthew/.chefdk/gem/ruby/2.1.0/gems/chef-provisioning-vsphere-0.5.7/lib/chef/provisioning/vsphere_driver/vsphere_helpers.rb:169:in `do_vm_clone'
C:/Users/Matthew/.chefdk/gem/ruby/2.1.0/gems/chef-provisioning-vsphere-0.5.7/lib/chef/provisioning/vsphere_driver/driver.rb:494:in `clone_vm'
C:/Users/Matthew/.chefdk/gem/ruby/2.1.0/gems/chef-provisioning-vsphere-0.5.7/lib/chef/provisioning/vsphere_driver/driver.rb:177:in `find_or_create_vm'
C:/Users/Matthew/.chefdk/gem/ruby/2.1.0/gems/chef-provisioning-vsphere-0.5.7/lib/chef/provisioning/vsphere_driver/driver.rb:135:in `allocate_machine'
C:/opscode/chefdk/embedded/lib/ruby/gems/2.1.0/gems/chef-provisioning-1.1.1/lib/chef/provider/machine.rb:30:in `block in <class:Machine>'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/provider/lwrp_base.rb:60:in `instance_eval'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/provider/lwrp_base.rb:60:in `recipe_eval_with_update_check'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/provider/lwrp_base.rb:45:in `block in action'
C:/opscode/chefdk/embedded/lib/ruby/gems/2.1.0/gems/chef-provisioning-1.1.1/lib/chef/provider/machine.rb:37:in `block in <class:Machine>'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/provider/lwrp_base.rb:60:in `instance_eval'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/provider/lwrp_base.rb:60:in `recipe_eval_with_update_check'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/provider/lwrp_base.rb:45:in `block in action'
C:/opscode/chefdk/embedded/lib/ruby/gems/2.1.0/gems/chef-provisioning-1.1.1/lib/chef/provider/machine.rb:55:in `block in <class:Machine>'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/provider/lwrp_base.rb:60:in `instance_eval'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/provider/lwrp_base.rb:60:in `recipe_eval_with_update_check'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/provider/lwrp_base.rb:45:in `block in action'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/provider.rb:130:in `run_action'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/resource.rb:562:in `run_action'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/runner.rb:49:in `run_action'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/runner.rb:81:in `block (2 levels) in converge'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/runner.rb:81:in `each'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/runner.rb:81:in `block in converge'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/resource_collection/resource_list.rb:83:in `block in execute_each_resource'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/resource_collection/stepable_iterator.rb:116:in `call'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/resource_collection/stepable_iterator.rb:116:in `call_iterator_block'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/resource_collection/stepable_iterator.rb:85:in `step'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/resource_collection/stepable_iterator.rb:104:in `iterate'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/resource_collection/stepable_iterator.rb:55:in `each_with_index'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/resource_collection/resource_list.rb:81:in `execute_each_resource'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/runner.rb:80:in `converge'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/client.rb:339:in `block in converge'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/client.rb:334:in `catch'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/client.rb:334:in `converge'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/client.rb:353:in `converge_and_save'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/client.rb:457:in `run'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/application.rb:253:in `run_with_graceful_exit_option'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/application.rb:230:in `block in run_chef_client'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/local_mode.rb:39:in `with_server_connectivity'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/application.rb:213:in `run_chef_client'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/application/client.rb:402:in `block in interval_run_chef_client'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/application/client.rb:392:in `loop'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/application/client.rb:392:in `interval_run_chef_client'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/application/client.rb:382:in `run_application'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/application.rb:60:in `run'
C:/opscode/chefdk/embedded/apps/chef/bin/chef-client:26:in `<top (required)>'
C:/opscode/chefdk/bin/chef-client:63:in `load'
C:/opscode/chefdk/bin/chef-client:63:in `<main>'

**Comment by mwrock** [comment]
Do you mind sharing the structure of your Hosts and Clusters view? It sounds to me like there is something in the driver code that searches for the actual host is not returning the `HostSystem` but rather a tree (`array`). The `host` option can take a path which begins at the root datacenter entity that you would see in the hosts and clusters view in the vmware client.

So if you had a cluster or a folder that contained your host, you would need to specify `folder_name/host_name` to the config option. If you do have folders that house your host make sure to include the complete path. If you still have issues, it would be helpful to know more about your structure unless your hosts hang directly off of the root datacenter.

**Comment by g0atcheez** [comment]
Matt,

Thanks for taking the time to respond. Indeed, my hosts are indeed just hanging off of root. I thought it could be that I wasn't specifying the path properly, but not sure of the format.

It looks like:

192.168.1.100
|
----- AD
|
----- Chef

Templates are stored in a separate folder.

**Comment by mwrock** [comment]
ok. Thanks for the info. Can you run this in an irb session:

```
require 'rbvmomi'
options = {
  host: '192.168.1.112',
  insecure: true,
  user:     '****',
  password: '****'
}
conn = RbVmomi::VIM.connect options
dc = conn.serviceInstance.find_datacenter('Home')
hosts = dc.hostFolder
host = hosts.childEntity.find { |f| f.name == '192.168.1.112' }

puts host.class
```

Of course insert your creds. 

I'd expect if your Hosts and Clusters view looked like:
```
Home
|-192.168.1.112
```
then the above code would result in `HostSystem`. However, based on the error, that must not be the case and the script may output simply `Array`. In that case I'd be curious to know what if anything is inside of the array.  So if it is just an array, could you add:

```
puts "host has #{host.count} items"
host.each {|h| puts h.class }
```
This info would really help to troubleshoot this.

**Comment by g0atcheez** [comment]
Looks like host.class is returning a ComputeResource object.

Here's the documentation on that from the vSphere web api: http://pubs.vmware.com/vsphere-60/index.jsp#com.vmware.wssdk.apiref.doc/vim.ComputeResource.html

I have one host and a VM running vCenter on it.

ESXi host: 192.168.1.100
vCenter server: 192.168.1.112

**Comment by mwrock** [comment]
Oh my apologies. Can you replace:
```
host = hosts.childEntity.find { |f| f.name == '192.168.1.112' }
```
with
```
host = hosts.childEntity.find { |f| f.name == '192.168.1.100' }
```

That was a oversight and explains why it returned a `ComputeResource`.

**Comment by g0atcheez** [comment]
I forgot to mention to you that whenever I ran the code in irb, I ran it with the ESXi host IP address and not the vCenter IP because I assumed that's what I needed to do.

It returns the ComputeResource instead of a HostSystem object.

**Comment by mwrock** [comment]
oh ok. Would it be possible to get a screen shot of your "Hosts and Clusters" view?

**Comment by g0atcheez** [comment]
![image](https://cloud.githubusercontent.com/assets/8755975/8129793/410dabd2-10d2-11e5-9739-48b3d95aac03.png)

Here ya go

**Comment by mwrock** [comment]
Thanks!

Can I trouble you to run one more script?

```
require 'rbvmomi'
options = {
  host: '192.168.1.112',
  insecure: true,
  user:     '****',
  password: '****'
}
conn = RbVmomi::VIM.connect options
dc = conn.serviceInstance.find_datacenter('Home')
hosts = dc.hostFolder
compute = hosts.childEntity.find { |f| f.name == '192.168.1.100' }
host = compute.host.first

puts host.class
```

**Comment by g0atcheez** [comment]
Nice! This time I got back a HostSystem object.

    irb(main):052:0> puts host.class
    HostSystem
    => nil

So I guess the question is now, how can I modify my provision class to get back this result instead of the ComputeResource object.

**Comment by mwrock** [comment]
Great! Thanks for your help getting to the bottom of this. There is really nothing that can be done with the code in its current state to get the HostSystem. However, I can push a prerelease before the end of the day with a fix for this.

**Comment by g0atcheez** [comment]
Awesome, thanks for the help on this and I'll be watching for the prerelease. Please keep up the good work on this gem. I looked at several options for provisioning to vCenter and none are as complete and well documented as this.

**Comment by mwrock** [comment]
just released 0.5.7.dev. let me know if you have issues. Thanks!

**Comment by g0atcheez** [comment]
Matt,

Unfortunately it looks like I'm still having issues. After installing the dev gem, I'm getting an undefined method 'first' error.

    [2015-06-12T23:39:35-05:00] FATAL: NoMethodError: machine[my_machine_name]  (vcenter_provisioning_test::provision line 36) had an error: NoMethodError: undefined method `first' for #<RbVmomi::VIM::ComputeResource:0x705bfd0>

**Comment by mwrock** [comment]
ok try grabbing 0.5.7.dev1

This is a bit tricky to get right since none of my VCenters seem to be configured quite like yours but I believe I see the issue preventing the .dev build from working for you.

**Comment by g0atcheez** [comment]
I feel like I'm getting further now, but now it's failing on actually trying to build the VM instead of finding the host.

    2015-06-13T10:04:55-05:00] FATAL: RbVmomi::Fault: machine[my_machine_name] (vcenter_provisioning_test::provision line 37) had an error: RbVmomi::Fault: InvalidArgument: A specified parameter was not correct: spec.location.pool

Also, I was getting an error about the product ID, but then I figured out that it's mandatory for a Windows machine to provide that in your cspec. You might want to put a blurb in the Windows specific stuff that indicates that.

Stack trace
> RbVmomi::Fault: machine[my_machine_name] (vcenter_provisioning_test::provision line 37) had an error: RbVmomi::Fault: InvalidArgument: A specified parameter was not correct: spec.location.pool
C:/Users/Matthew/.chefdk/gem/ruby/2.1.0/gems/rbvmomi-1.8.2/lib/rbvmomi/vim/Task.rb:11:in `wait_for_completion'
C:/Users/Matthew/.chefdk/gem/ruby/2.1.0/gems/chef-provisioning-vsphere-0.5.7.dev1/lib/chef/provisioning/vsphere_driver/driver.rb:504:in `clone_vm'
C:/Users/Matthew/.chefdk/gem/ruby/2.1.0/gems/chef-provisioning-vsphere-0.5.7.dev1/lib/chef/provisioning/vsphere_driver/driver.rb:177:in `find_or_create_vm'
C:/Users/Matthew/.chefdk/gem/ruby/2.1.0/gems/chef-provisioning-vsphere-0.5.7.dev1/lib/chef/provisioning/vsphere_driver/driver.rb:135:in `allocate_machine'
C:/opscode/chefdk/embedded/lib/ruby/gems/2.1.0/gems/chef-provisioning-1.1.1/lib/chef/provider/machine.rb:30:in `block in <class:Machine>'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/provider/lwrp_base.rb:60:in `instance_eval'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/provider/lwrp_base.rb:60:in `recipe_eval_with_update_check'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/provider/lwrp_base.rb:45:in `block in action'
C:/opscode/chefdk/embedded/lib/ruby/gems/2.1.0/gems/chef-provisioning-1.1.1/lib/chef/provider/machine.rb:37:in `block in <class:Machine>'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/provider/lwrp_base.rb:60:in `instance_eval'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/provider/lwrp_base.rb:60:in `recipe_eval_with_update_check'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/provider/lwrp_base.rb:45:in `block in action'
C:/opscode/chefdk/embedded/lib/ruby/gems/2.1.0/gems/chef-provisioning-1.1.1/lib/chef/provider/machine.rb:55:in `block in <class:Machine>'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/provider/lwrp_base.rb:60:in `instance_eval'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/provider/lwrp_base.rb:60:in `recipe_eval_with_update_check'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/provider/lwrp_base.rb:45:in `block in action'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/provider.rb:130:in `run_action'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/resource.rb:562:in `run_action'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/runner.rb:49:in `run_action'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/runner.rb:81:in `block (2 levels) in converge'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/runner.rb:81:in `each'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/runner.rb:81:in `block in converge'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/resource_collection/resource_list.rb:83:in `block in execute_each_resource'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/resource_collection/stepable_iterator.rb:116:in `call'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/resource_collection/stepable_iterator.rb:116:in `call_iterator_block'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/resource_collection/stepable_iterator.rb:85:in `step'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/resource_collection/stepable_iterator.rb:104:in `iterate'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/resource_collection/stepable_iterator.rb:55:in `each_with_index'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/resource_collection/resource_list.rb:81:in `execute_each_resource'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/runner.rb:80:in `converge'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/client.rb:339:in `block in converge'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/client.rb:334:in `catch'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/client.rb:334:in `converge'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/client.rb:353:in `converge_and_save'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/client.rb:457:in `run'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/application.rb:253:in `run_with_graceful_exit_option'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/application.rb:230:in `block in run_chef_client'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/local_mode.rb:39:in `with_server_connectivity'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/application.rb:213:in `run_chef_client'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/application/client.rb:402:in `block in interval_run_chef_client'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/application/client.rb:392:in `loop'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/application/client.rb:392:in `interval_run_chef_client'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/application/client.rb:382:in `run_application'
C:/opscode/chefdk/embedded/apps/chef/lib/chef/application.rb:60:in `run'
C:/opscode/chefdk/embedded/apps/chef/bin/chef-client:26:in `<top (required)>'
C:/opscode/chefdk/bin/chef-client:63:in `load'
C:/opscode/chefdk/bin/chef-client:63:in `<main>'

Seems like it is trying to use a resource pool even though I didn't specify one in my provision recipe.

**Comment by mwrock** [comment]
Well I'm glad we are past the host problem. The resource pool issue is likely the result of recent refactoring in building the clone spec to make it testable. So opportunity for more tests:). 

I think it should be easy to make the product is optional. And its probably the right thing to do.

Will let you know when we have a new dev gem to test.

Thanks!

**Comment by mwrock** [comment]
There is a 0.5.7.dev2 now that will spit out a string representation of the vmware clone_spec prior to cloning. Can you run with debug verbosity and provide the clone_spec output? Of course ommiting any sensitive data.

Looking at the code, if you pass a host, we are not passing a resource pool. At least not explicitly.  So it would be helpful to see what is in the spec.

**Comment by g0atcheez** [comment]
I'm assuming you want to see the cspec that's generated and sent over to the vSphere api?

Since it's quite long I put it in a pastbin: http://pastebin.com/AJcD0Sfh

Let me know if you need other info.

**Comment by mwrock** [comment]
Lookin over the VMware docs: https://www.vmware.com/support/developer/converter-sdk/conv43_apireference/vim.vm.RelocateSpec.html

Looks like a resource pool is required when cloning from a template to a vm. We don't clone from templates which must be why I've not run into this before.

Currently the driver will use either the host you specify or the resource pool prefering the host but it won't use both.

Sounds like you will need to pass a pool but until we update this you will need to omit host. However in your case if you just have one host, specifying it shouldn't be necessary with the pool.

Could you try creating a pool and passing that instead of host?

**Comment by g0atcheez** [comment]
I created two resource pools and would like to place my provisioned VM into the 'Test' resource pool.

![image](https://cloud.githubusercontent.com/assets/8755975/8146129/0c46c970-11ee-11e5-8d05-9d180e3b4e75.png)

How do I need to name the pool in my provision script? I tried just using Test and it's saying it can't find the resource pool. I don't have any clusters setup.

    [2015-06-13T16:59:09-05:00] FATAL: RuntimeError: machine[my_machine_name] (vcenter_provisioning_test::provision line 37) had an error: RuntimeError: vSphere ResourcePool not found [Test]

**Comment by mwrock** [comment]
try '192.168.1.100/Test'

**Comment by g0atcheez** [comment]
Same message.

    [2015-06-14T08:28:42-05:00] FATAL: RuntimeError: machine[my_machine_name] (vcenter_provisioning_test::provision line 37) had an error: RuntimeError: vSphere ResourcePool not found [192.168.1.100/Test]

Any ideas?

**Comment by mwrock** [comment]
Lets start troubleshooting by running this in irb.
```
require 'rbvmomi'
options = {
  host: '192.168.1.112',
  insecure: true,
  user:     '****',
  password: '****'
}
conn = RbVmomi::VIM.connect options
dc = conn.serviceInstance.find_datacenter('Home')
hosts = dc.hostFolder
compute = hosts.childEntity.find { |f| f.name == '192.168.1.100' }
pool_root = compute.resourcePool
puts "root has #{pool_root.resourcePool.count} pools"
pool_root.resourcePool.each { |p| puts p.name }
```

I'd expect this to output:
```
root has 2 pools
Infra
Test
```
However, if that were the case then the driver should have found it.

**Comment by g0atcheez** [comment]
Yep, that's exactly what the output is.

    root has 2 pools
    irb(main):040:0> pool_root.resourcePool.each { |p| puts p.name }
    Test
    Infra

**Comment by mwrock** [comment]
ok weird. I'll drop another dev gem with more debug output

**Comment by mwrock** [comment]
give 0.5.7.dev3 a try and provide a link to the debugging. Thanks!

**Comment by g0atcheez** [comment]
First time I did it with 'Test' as the resource pool name and got the expected not found error. Trying with '192.168.1.100/Test' yields different results and it seems to get passed finding the resource pool but fails on what looks to be a typo somewhere.

    [2015-06-14T14:59:37-05:00] DEBUG: Finding pool: 192.168.1.100/Test
    [2015-06-14T14:59:37-05:00] DEBUG: Identifying entity part: 192.168.1.100 in folder type: Folder
    [2015-06-14T14:59:37-05:00] DEBUG: Parent folder is a folder
    [2015-06-14T14:59:37-05:00] DEBUG: Identifying entity part: Test in folder type: ComputeResource

    [2015-06-14T14:59:38-05:00] FATAL: NameError: machine[my_machine_name] (vcenter_provisioning_test::provision line 37) had an error: NameError: undefined local variable or method `patrent' for #<ChefProvisioningVsphere::VsphereHelper:0x6f1aac0>

**Comment by mwrock** [comment]
ok. typo in logging but enough to show me the problem. Hope to have a fix soon to at least get past this layer.

**Comment by mwrock** [comment]
ok try 0.5.7.dev4

**Comment by g0atcheez** [comment]
It's not liking my provision recipe. It's telling me to specify a host or resource_pool and I'm specifying a pool but it's not working.
````ruby
    with_machine_options :bootstrap_options => {
      num_cpus: 1,
      memory_mb: 4096,
      network_name: ['VM Network'],
      datacenter: 'Home',
      resource_pool: '192.168.1.100/Test',
      #host: '192.168.1.100',
      datastore: 'esxlab_storage',
      template_name: 'WIN2008R2PS4',
      customization_spec: {
         ipsettings: {
         dnsServerList: ['192.168.1.110','192.168.1.1']
        },
        domain: 'casa.local',
        domainAdmin: 'administrator@casa.local',
        domainAdminPassword: '****',
        product_id: '****',
        org_name: 'Casa',
        win_time_zone: 20
      },
    ssh: {
        user: 'Administrator',
        password: '****',
        paranoid: false,
    }
    },
    :convergence_options => {
        :ssl_verify_mode => :verify_none
    }
````
    [2015-06-14T15:44:01-05:00] FATAL: RuntimeError: machine[my_machine_name] (vcenter_provisioning_test::provision line 38) had an error: RuntimeError: either :host or :resource_pool must be specified           when cloning from a VM Template

**Comment by mwrock** [comment]
ok. lets see where 0.5.7.dev5 gets you.

I apologize for the back and forth here but its non trivial for me to setup a test vcenter matching your setup. Really appreciate your patience.

**Comment by g0atcheez** [comment]
Success!! I still have some work to be done on my template, but it's creating the VM now and trying to run through the bootstrap process so provisioning is working. Thank you SO much for your assistance on this, it was very helpful.

**Comment by mwrock** [comment]
YAY! Thank YOU @g0atcheez!

**Comment by mwrock** [comment]
Just an FYI that I just released another dev gem 0.5.7.dev6 that will allow you to not need to create "dummy" pools. You could specify the host like you were doing before and if you are using a vm template and do not specify a resource pool, the driver will simply assign to the resource pool "root" of your host. This is basically a hidden resource pool that always exists under every cluster or host.

I've tested under our infrastructure but we use clusters exclusively so its always possible you could hit a snag but I think this should behave even without clusters.

Let me know if you try it and run into any troubles.
